src/main/software/amazon/event/ruler/ByteMap.java

```diff
@@ -193,7 +191,13 @@ ByteTransition getTransitionForAllBytes() {
      * @return All transitions contained in this map.
      */
     Set<ByteTransition> getTransitions() {
-        return map.values().stream().filter(Objects::nonNull).collect(Collectors.toSet());
+        Set<ByteTransition> result = new HashSet<>(map.values().size());
```

Comment:
Just a note that the Set may end up containing less elements than map.values().size(), because 1) nulls are removed, and 2) the values in the map are not unique. I suppose it's probably more efficient to oversize the Set initially though than to undersize it by going with the default constructor.

Reply:
yes good observation, I was aware of this, I just took the size of the values as the upper bound of the set size, but perf wise its cheaper to allocate above, than allowing the set to grow